core/src/main/scala/com/mesosphere/usi/core/matching/RangeResource.scala

```diff
@@ -0,0 +1,251 @@
+import scala.util.Random
+
+/**
+  * Represents requirement for Mesos resource of type Range.
+  * You can either request static values (by providing concrete value) or dynamic values (by providing 0).
+  * Dynamic values are then selected from the offered ranges. If random implementation is passed in, these ranges are randomized.
+  * For more information about randomization see [[lazyRandomValuesFromRanges()]]
+  *
+  * @param requestedValues values pod wants to consume on the given ranges
+  * @param resourceType name of resource (e.g. ports)
+  * @param random when requesting dynamic values, you can provide Random implementation if you want dynamic values to be randomized
+  */
+case class RangeResource(requestedValues: Seq[Int], resourceType: ResourceType, random: Random = Random)
+    extends ResourceRequirement {
+  override def description: String = s"$resourceType:[${requestedValues.mkString(",")}]"
+
+  override def matchAndConsume(resources: Seq[Protos.Resource]): Option[ResourceMatchResult] = {
+    matchAndConsumeIter(Nil, resources.toList)
+  }
+
+  /**
+    * Iterates over given resources keeping in track unmatched resources.
+    * In the end produces ResourceMatchResult if we were able to find a match on the given resources.
+    *
+    * @param unmatchedResources already processed resources that did not contain resources we are trying to match
+    * @param remainingResources resources we still need to process when looking for resource match
+    * @return None if no match was possible, ResourceMatchResult if we were able to consume requested resources
+    */
+  @tailrec private def matchAndConsumeIter(
+      unmatchedResources: List[Protos.Resource],
+      remainingResources: List[Protos.Resource]): Option[ResourceMatchResult] = {
+
+    remainingResources match {
+      case Nil =>
+        None
+      case next :: rest =>
+        tryConsumeValuesFromResource(requestedValues, next) match {
+          case Nil =>
+            matchAndConsumeIter(next :: unmatchedResources, rest)
+          case consumedResources =>
+            Some(
+              ResourceMatchResult(
+                consumedResources,
+                unmatchedResources ++ rest ++ ResourceUtil.consumeResources(Seq(next), consumedResources)))
+        }
+    }
+  }
+
+  /**
+    * Tries to match and consume values from a given resource
+    * @param values values we want to consume
+    * @param resource mesos resource to match agains
+    * @return final list of mesos resources created after consuming requested values, empty if match was not possible
+    */
+  private def tryConsumeValuesFromResource(values: Seq[Int], resource: Protos.Resource): Seq[Protos.Resource] = {
+    val offeredRanges = parseResourceToRanges(resource)
+    if (offeredRanges.isEmpty || values.isEmpty) {
+      return Seq.empty
+    }
+
+    // non-dynamic values
+    val staticRequestedValues = values.collect { case v if v != 0 => v }.toSet
```

Comment:
maybe just `values.filter(_ != 0)`?

Reply:
haha yeah, that's probably the Scala you were talking about yesterday :)